feat: add beforeEach and afterEach hooks

hyanmandian opened the pull request:
Closes #9

lukeed commented:
Hey thanks!

This still has some work to go (types and docs), but we also need to thinking about the implementation. I've had this originally too, but it doesn't account for:

* `beforeEach` throwing an Error
* `afterEach` throwing an Error
* `test` throwing an Error & so `afterEach` can't cleanup

I will also ping you from #9 about another discussion point

hyanmandian commented:
I implemented the idea you suggested in #9 , added missing docs, added missing types, and handle hooks exceptions. Can you help me to find a way to test if it works?

lukeed commented:
Sorry @hyanmandian 🙈 Can you revert that last commit? It actually causes more issues with timing and control flow than the previous state.

The rest is looking good! I can add the missing bits to what _was_ there.
Thanks!

hyanmandian commented:
Done @lukeed !

lukeed commented:
This is looking good, thanks @hyanmandian! I'll be able to add the remaining bits tonight 🙌

lukeed commented on src/index.js:
```diff
@@ -67,10 +66,13 @@ async function runner(ctx, name) {
 				write(PASS);
 				num++;
 			} catch (err) {
+				for (hook of aEach) await hook();
```
Sorry for the delay – I was thinking about it, and this is all the "additional cleanup" that was needed.

hyanmandian replied:
Nice!
